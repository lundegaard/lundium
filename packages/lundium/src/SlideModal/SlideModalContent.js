import React from 'react';
import PropTypes from 'prop-types';
import { cx } from 'ramda-extension';

import Box from '../Box';

const SlideModalContent = ({ className, ...otherProps }) => (
	<Box className={cx('slide-modal__content', className)} {...otherProps} />
);

SlideModalContent.propTypes = {
	/** Class name for root. */
	className: PropTypes.string,
};

export default SlideModalContent;
