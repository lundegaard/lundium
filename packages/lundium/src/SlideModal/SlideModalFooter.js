import React from 'react';
import PropTypes from 'prop-types';
import { cx } from 'ramda-extension';

import Box from '../Box';

const SlideModalFooter = ({ className, ...otherProps }) => (
	<Box className={cx('slide-modal__footer', className)} {...otherProps} />
);

SlideModalFooter.propTypes = {
	/** Class name for root. */
	className: PropTypes.string,
};

export default SlideModalFooter;
