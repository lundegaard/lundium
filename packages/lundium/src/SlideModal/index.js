export { default } from './SlideModal';
export { default as BasicSlideModal } from './BasicSlideModal';

export { default as SlideModalFooter } from './SlideModalFooter';
export { default as SlideModalHeader } from './SlideModalHeader';
export { default as SlideModalContent } from './SlideModalContent';
