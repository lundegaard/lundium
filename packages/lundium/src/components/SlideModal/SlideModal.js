import React, { useEffect } from 'react';
import { cx, noop } from 'ramda-extension';
import PropTypes from 'prop-types';

import withTransition from '../../utils/withTransition';
import stopPropagation from '../../utils/stopPropagation';
import Box from '../Box';
import Grid from '../Grid';
import Button from '../Button';
import Icon from '../Icon';
import PortalWithBackdrop from '../PortalWithBackdrop';

export const SlideModal = ({
	isVisible,
	className,
	children,
	onClose = noop,
	onOpen = noop,
	size = 'sm',
	...otherProps
}) => {
	useEffect(() => {
		if (isVisible) {
			onOpen();
		}
	}, [isVisible, onOpen]);

	return (
		<PortalWithBackdrop isVisible={isVisible} onBackdropClick={onClose}>
			<Box
				className={cx('slide-modal', `slide-modal--${size}`, className)}
				onClick={stopPropagation}
				{...otherProps}
			>
				<Grid
					justifyContent="end"
					alignItems="center"
					className="slide-modal__controls"
				>
					<Button kind="blank" size="xs" onClick={onClose}>
						<Icon type="cancel" size={18} />
					</Button>
				</Grid>
				{children}
			</Box>
		</PortalWithBackdrop>
	);
};

SlideModal.propTypes = {
	/** Child content for modal window */
	children: PropTypes.any,
	/** Root class name */
	className: PropTypes.string,
	/** Is modal visible */
	isVisible: PropTypes.bool,
	/** On cancel function */
	onClose: PropTypes.func,
	/** On handle open event */
	onOpen: PropTypes.func,
	/** Size (width) */
	size: PropTypes.oneOf(['sm', 'md', 'lg']),
};

export default withTransition(SlideModal, {
	trigger: 'isVisible',
});
