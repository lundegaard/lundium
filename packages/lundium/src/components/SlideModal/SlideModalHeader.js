import React from 'react';
import PropTypes from 'prop-types';
import { cx } from 'ramda-extension';

import Grid from '../Grid';
import Text from '../Text';

const SlideModalHeader = ({
	children,
	className,
	title,
	subtitle,
	details,
	...otherProps
}) => (
	<Grid
		justifyContent="between"
		alignItems="end"
		flexWrap
		className={cx('slide-modal__header', className)}
		{...otherProps}
	>
		<Text className="slide-modal__title">{title}</Text>
		{subtitle && <Text className="slide-modal__sub-title">{subtitle}</Text>}
		{details && (
			<Grid
				flex
				flexWrap
				flexDirection={['column', 'row']}
				className="slide-modal__details"
			>
				{details}
			</Grid>
		)}
		{children}
	</Grid>
);

SlideModalHeader.propTypes = {
	/** Children to be rendered in the main container. */
	children: PropTypes.node,
	/** Class name for root. */
	className: PropTypes.string,
	details: PropTypes.node,
	subtitle: PropTypes.node,
	title: PropTypes.node.isRequired,
};

export default SlideModalHeader;
