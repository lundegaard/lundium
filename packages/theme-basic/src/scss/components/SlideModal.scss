@import '../core';

/*
 -  Slide Modal
 -  (<div class="slide-modal" />)
 */


.slide-modal {
	position: fixed;
	inset-inline-end: 0;
	inset-block-start: 0;
	height: 100vh;
	width: 100%;
	max-width: $slide-modal-width;
	background-color: $slide-modal-bg;
	box-shadow: $box-shadow-basic;
	overflow-y: auto;
	z-index: $zindex-modal;

	// Media queries
	@include media-breakpoint-up(md) {
		width: 70%;
	}

	&--sm {
		// Media queries
		@include media-breakpoint-up(md) {
			width: $slide-modal-width-sm;

			.slide-modal__footer {
				width: $slide-modal-width-sm;
			}
		}
	}

	&--lg {
		// Media queries
		@include media-breakpoint-up(md) {
			width: 100%;
		}
		@include media-breakpoint-up(lg) {
			width: auto;
			max-width: none;
			inset-inline-start: $main-nav-width;

			.slide-modal__footer {
				width: auto;
				max-width: none;
				inset-inline-start: $main-nav-width;
			}
		}
	}

	&__controls {
		position: absolute;
		inset-block-start: 0;
		inset-inline: 0;
		height: $slide-modal-controls-mobile-height;
		padding: logical $slide-modal-mobile-padding-y $slide-modal-mobile-padding-x 0;

		// Media queries
		@include media-breakpoint-up(md) {
			height: $slide-modal-controls-height;
			padding: logical $slide-modal-padding-y $slide-modal-padding-x 0;
		}
	}

	&__header {
		margin-block: ( $spacer * .5 ) ( $spacer * 1.5 );
		padding: logical $slide-modal-controls-mobile-height $slide-modal-mobile-padding-x 0;

		// Media queries
		@include media-breakpoint-up(md) {
			padding: logical $slide-modal-controls-height $slide-modal-padding-x 0 ;
		}
	}

	&__title {
		width: 100%;
		font-weight: 800;
		font-size: 2.6rem;
		margin-block-end: ( $spacer * .5 );
		line-height: 1.2;

		// Media queries
		@include media-breakpoint-up(sm) {
			width: auto;
			margin-block-end: 0;
			font-size: 3.6rem;
		}
	}

	&__sub-title {
		font-size: 1.8rem;
		margin-block: ( $spacer * .5 ) 0;
		@include text-truncate;

		// Media queries
		@include media-breakpoint-up(md) {
			font-size: 2rem;
		}
	}

	&__inner {
		display: flex;
		flex-direction: column;
		min-height: 100vh;
	}

	&__content {
		padding-inline: $slide-modal-mobile-padding-x;
		padding-block-end: calc(#{$slide-modal-footer-mobile-height} + #{$spacer});

		// Media queries
		@include media-breakpoint-up(md) {
			padding-inline: $slide-modal-padding-x;
		}
		@include media-breakpoint-up(xl) {
			padding-block-end: calc(#{$slide-modal-footer-height} + #{$spacer * 2});
		}

		.content-divider {
			margin-inline: -$grid-gutter-width;
		}

		.content-inner {
			padding-inline: 0;
		}
	}

	&__footer {
		width: 100%;
		margin-top: auto;
		height: $slide-modal-footer-mobile-height;
		padding: logical $slide-modal-mobile-padding-y $slide-modal-mobile-padding-x ( $slide-modal-mobile-padding-y * 2 );
		background-color: $slide-modal-bg;
		box-shadow: 0 -20px 40px 0 rgba($theme-black, .04);

		// Media queries
		@include media-breakpoint-up(md) {
			height: $slide-modal-footer-height;
			padding: logical $slide-modal-padding-y $slide-modal-padding-x;
		}
		@include media-breakpoint-up(xl) {
			position: fixed;
			inset-block-end: 0;
			inset-inline-end: 0;
			max-width: $slide-modal-width;
		}
	}
}

.slide-modal-transition {
	&-enter {
		.slide-modal {
			transform: translateX(100%);
		}

		&.backdrop {
			opacity: 0;
		}

		&-active {
			.slide-modal {
				transform: translateX(0%);
				transition: transform 1000ms ease-in-out;
			}

			&.backdrop {
				opacity: 1;
			}
		}
	}

	&-exit {
		.slide-modal {
			transform: translateX(0%);
		}

		&.backdrop {
			opacity: 1;
		}

		&-active {
			.slide-modal {
				transform: translateX(100%);
				transition: transform 1000ms ease-in-out;
			}

			&.backdrop {
				opacity: 0;
			}
		}
	}
}
