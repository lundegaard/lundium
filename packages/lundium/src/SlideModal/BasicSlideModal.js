import React from 'react';
import PropTypes from 'prop-types';

import SlideModalFooter from './SlideModalFooter';
import SlideModalHeader from './SlideModalHeader';
import SlideModalContent from './SlideModalContent';
import SlideModal from './SlideModal';

export const BasicSlideModal = ({
	renderFooter: Footer,
	children,
	title,
	subtitle,
	details,
	...otherProps
}) => (
	<SlideModal {...otherProps}>
		{title && (
			<SlideModalHeader title={title} subtitle={subtitle} details={details} />
		)}
		<SlideModalContent>{children}</SlideModalContent>
		{Footer && (
			<SlideModalFooter>
				<Footer />
			</SlideModalFooter>
		)}
	</SlideModal>
);

BasicSlideModal.propTypes = {
	children: PropTypes.node,
	details: PropTypes.node,
	renderFooter: PropTypes.elementType,
	subtitle: PropTypes.node,
	title: PropTypes.node,
};

export default BasicSlideModal;
