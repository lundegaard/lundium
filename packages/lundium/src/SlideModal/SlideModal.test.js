import React from 'react';
import { shallow } from 'enzyme';

import { SlideModal } from './SlideModal';

describe('<SlideModal />', () => {
	const requiredProps = { children: <div>Some child</div> };

	it('matches snapshot when shallowly rendered', () => {
		const wrapper = shallow(<SlideModal {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});
