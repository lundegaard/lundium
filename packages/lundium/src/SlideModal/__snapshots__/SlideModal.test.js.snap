// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlideModal /> matches snapshot when shallowly rendered 1`] = `
<PortalWithBackdrop
  onBackdropClick={[Function]}
>
  <forwardRef(Box)
    className="slide-modal slide-modal--sm"
    onClick={[Function]}
  >
    <forwardRef(Grid)
      alignItems="center"
      className="slide-modal__controls"
      justifyContent="end"
    >
      <ForwardRef
        kind="blank"
        onClick={[Function]}
        size="xs"
      >
        <Icon
          size={18}
          type="cancel"
        />
      </ForwardRef>
    </forwardRef(Grid)>
    <div>
      Some child
    </div>
  </forwardRef(Box)>
</PortalWithBackdrop>
`;
